File: stage2-gui/python-backend/key_detector.py
"""
Musical key detection module using chromagram analysis.
Implements Krumhansl-Schmuckler key-finding algorithm.
"""

import numpy as np
import librosa
from typing import Tuple


class KeyDetector:
    """Detects musical key from audio data."""

    # Krumhansl-Schmuckler key profiles
    # Major key profile (correlation weights for each pitch class)
    MAJOR_PROFILE = np.array([6.35, 2.23, 3.48, 2.33, 4.38, 4.09, 2.52, 5.19, 2.39, 3.66, 2.29, 2.88])

    # Minor key profile
    MINOR_PROFILE = np.array([6.33, 2.68, 3.52, 5.38, 2.60, 3.53, 2.54, 4.75, 3.98, 2.69, 3.34, 3.17])

    # Pitch class names
    PITCH_CLASSES = ['C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B']

    def __init__(self, sample_rate: int = 22050):
        """
        Initialize the key detector.

        Args:
            sample_rate: Sample rate of the audio data
        """
        self.sample_rate = sample_rate

    def detect(self, audio_data: np.ndarray) -> Tuple[str, str, float]:
        """
        Detect the musical key of the audio.

        Args:
            audio_data: Audio data as numpy array

        Returns:
            Tuple of (key, mode, confidence)
            - key: Root note (C, C#, D, etc.)
            - mode: "major" or "minor"
            - confidence: Confidence score (0-1)

        Raises:
            ValueError: If audio data is invalid
        """
        if audio_data is None or len(audio_data) == 0:
            raise ValueError("Audio data is empty")

        if len(audio_data) < self.sample_rate:
            raise ValueError("Audio data is too short (minimum 1 second)")

        try:
            # Compute chromagram
            chroma = librosa.feature.chroma_cqt(
                y=audio_data,
                sr=self.sample_rate,
                hop_length=512
            )

            # Average across time to get pitch class distribution
            pitch_class_distribution = np.mean(chroma, axis=1)

            # Normalize the distribution
            pitch_class_distribution = pitch_class_distribution / np.sum(pitch_class_distribution)

            # Find best matching key using Krumhansl-Schmuckler algorithm
            key, mode, confidence = self._find_best_key(pitch_class_distribution)

            return key, mode, confidence

        except Exception as e:
            raise RuntimeError(f"Error detecting key: {str(e)}") from e

    def _find_best_key(self, pitch_distribution: np.ndarray) -> Tuple[str, str, float]:
        """
        Find the best matching key using correlation with key profiles.

        Args:
            pitch_distribution: Normalized pitch class distribution

        Returns:
            Tuple of (key, mode, confidence)
        """
        max_correlation = -1
        best_key = 'C'
        best_mode = 'major'

        # Try all 12 keys in both major and minor
        for i in range(12):
            # Rotate the profiles to match each key
            major_profile = np.roll(self.MAJOR_PROFILE, i)
            minor_profile = np.roll(self.MINOR_PROFILE, i)

            # Calculate correlation with major profile
            major_corr = self._correlation(pitch_distribution, major_profile)

            # Calculate correlation with minor profile
            minor_corr = self._correlation(pitch_distribution, minor_profile)

            # Check if major key is the best so far
            if major_corr > max_correlation:
                max_correlation = major_corr
                best_key = self.PITCH_CLASSES[i]
                best_mode = 'major'

            # Check if minor key is the best so far
            if minor_corr > max_correlation:
                max_correlation = minor_corr
                best_key = self.PITCH_CLASSES[i]
                best_mode = 'minor'

        # Normalize correlation to 0-1 confidence range
        confidence = (max_correlation + 1) / 2  # Correlation ranges from -1 to 1

        return best_key, best_mode, confidence

    def _correlation(self, x: np.ndarray, y: np.ndarray) -> float:
        """
        Calculate Pearson correlation coefficient.

        Args:
            x: First array
            y: Second array

        Returns:
            Correlation coefficient
        """
        # Normalize arrays
        x_normalized = (x - np.mean(x)) / (np.std(x) + 1e-10)
        y_normalized = (y - np.mean(y)) / (np.std(y) + 1e-10)

        # Calculate correlation
        correlation = np.mean(x_normalized * y_normalized)

        return correlation

    def get_key_string(self, key: str, mode: str) -> str:
        """
        Format key and mode as a readable string.

        Args:
            key: Root note
            mode: "major" or "minor"

        Returns:
            Formatted key string (e.g., "C major", "A minor")
        """
        return f"{key} {mode}"

    def get_confidence_level(self, confidence: float) -> str:
        """
        Convert confidence score to a human-readable level.

        Args:
            confidence: Confidence score (0-1)

        Returns:
            Confidence level string: "Low", "Medium", or "High"
        """
        if confidence >= 0.7:
            return "High"
        elif confidence >= 0.5:
            return "Medium"
        else:
            return "Low"

    def get_relative_keys(self, key: str, mode: str) -> dict:
        """
        Get related keys (relative, parallel, dominant).

        Args:
            key: Root note
            mode: "major" or "minor"

        Returns:
            Dictionary with related keys
        """
        key_idx = self.PITCH_CLASSES.index(key)

        result = {}

        if mode == 'major':
            # Relative minor (6 semitones down)
            relative_idx = (key_idx - 3) % 12
            result['relative'] = f"{self.PITCH_CLASSES[relative_idx]} minor"

            # Parallel minor (same root)
            result['parallel'] = f"{key} minor"

            # Dominant (5 semitones up)
            dominant_idx = (key_idx + 7) % 12
            result['dominant'] = f"{self.PITCH_CLASSES[dominant_idx]} major"

        else:  # minor
            # Relative major (3 semitones up)
            relative_idx = (key_idx + 3) % 12
            result['relative'] = f"{self.PITCH_CLASSES[relative_idx]} major"

            # Parallel major (same root)
            result['parallel'] = f"{key} major"

            # Dominant (7 semitones up)
            dominant_idx = (key_idx + 7) % 12
            result['dominant'] = f"{self.PITCH_CLASSES[dominant_idx]} minor"

        return result

    def get_scale_notes(self, key: str, mode: str) -> list:
        """
        Get the notes in the scale for the given key.

        Args:
            key: Root note
            mode: "major" or "minor"

        Returns:
            List of notes in the scale
        """
        key_idx = self.PITCH_CLASSES.index(key)

        if mode == 'major':
            # Major scale intervals: W W H W W W H (2 2 1 2 2 2 1)
            intervals = [0, 2, 4, 5, 7, 9, 11]
        else:  # natural minor
            # Natural minor scale intervals: W H W W H W W (2 1 2 2 1 2 2)
            intervals = [0, 2, 3, 5, 7, 8, 10]

        scale = [self.PITCH_CLASSES[(key_idx + interval) % 12] for interval in intervals]

        return scale

File: stage3-plugin/Source/KeyDetector.cpp
/*
  ==============================================================================

    Musical Key Detection Implementation
    Chromagram-based key detection using Krumhansl-Schmuckler algorithm

  ==============================================================================
*/

#include "KeyDetector.h"

void KeyDetector::prepare(double sr)
{
    sampleRate = sr;
}

std::tuple<juce::String, juce::String, float> KeyDetector::detectKey(
    const float* audioData, int numSamples)
{
    if (numSamples < fftSize)
        return {"C", "major", 0.0f};

    // Calculate chromagram (pitch class distribution)
    std::array<float, 12> chromagram = {0};
    calculateChromagram(audioData, numSamples, chromagram);

    // Normalize chromagram
    float sum = 0.0f;
    for (float val : chromagram)
        sum += val;

    if (sum > 0.0f)
    {
        for (float& val : chromagram)
            val /= sum;
    }

    // Find best matching key
    return findBestKey(chromagram);
}

void KeyDetector::calculateChromagram(const float* audioData, int numSamples,
                                     std::array<float, 12>& chromagram)
{
    chromagram.fill(0.0f);

    juce::dsp::FFT fft(12); // 4096 point FFT
    std::vector<std::complex<float>> fftData(fftSize);

    int numFrames = (numSamples - fftSize) / hopSize;

    for (int frame = 0; frame < numFrames; ++frame)
    {
        int startSample = frame * hopSize;

        // Apply Hann window
        std::vector<float> windowed(fftSize);
        for (int i = 0; i < fftSize; ++i)
        {
            float window = 0.5f * (1.0f - std::cos(2.0f * juce::MathConstants<float>::pi * i / (fftSize - 1)));
            windowed[i] = audioData[startSample + i] * window;
        }

        // Perform FFT
        fft.perform(windowed.data(), reinterpret_cast<float*>(fftData.data()), false);

        // Map FFT bins to pitch classes
        for (int bin = 1; bin < fftSize / 2; ++bin)
        {
            float frequency = bin * static_cast<float>(sampleRate) / fftSize;

            // Skip frequencies outside musical range (roughly 27.5 Hz to 4186 Hz)
            if (frequency < 27.5f || frequency > 4186.0f)
                continue;

            // Get magnitude
            float real = fftData[bin].real();
            float imag = fftData[bin].imag();
            float magnitude = std::sqrt(real * real + imag * imag);

            // Map to pitch class (0-11)
            float pitchClass = frequencyToPitchClass(frequency);
            int pitchClassIndex = static_cast<int>(pitchClass) % 12;

            chromagram[pitchClassIndex] += magnitude;
        }
    }
}

std::tuple<juce::String, juce::String, float> KeyDetector::findBestKey(
    const std::array<float, 12>& pitchClassDistribution)
{
    float maxCorrelation = -1.0f;
    juce::String bestKey = "C";
    juce::String bestMode = "major";

    // Try all 12 keys in both major and minor
    for (int root = 0; root < 12; ++root)
    {
        // Rotate profiles to match this root
        std::array<float, 12> rotatedMajor;
        std::array<float, 12> rotatedMinor;

        for (int i = 0; i < 12; ++i)
        {
            rotatedMajor[i] = majorProfile[(i + root) % 12];
            rotatedMinor[i] = minorProfile[(i + root) % 12];
        }

        // Calculate correlations
        float majorCorr = correlation(pitchClassDistribution, rotatedMajor);
        float minorCorr = correlation(pitchClassDistribution, rotatedMinor);

        // Check if major key is best so far
        if (majorCorr > maxCorrelation)
        {
            maxCorrelation = majorCorr;
            bestKey = pitchClasses[root];
            bestMode = "major";
        }

        // Check if minor key is best so far
        if (minorCorr > maxCorrelation)
        {
            maxCorrelation = minorCorr;
            bestKey = pitchClasses[root];
            bestMode = "minor";
        }
    }

    // Normalize correlation to 0-1 confidence
    float confidence = (maxCorrelation + 1.0f) / 2.0f;
    confidence = juce::jlimit(0.0f, 1.0f, confidence);

    return {bestKey, bestMode, confidence};
}

float KeyDetector::correlation(const std::array<float, 12>& x, const std::array<float, 12>& y)
{
    // Calculate means
    float meanX = 0.0f, meanY = 0.0f;
    for (int i = 0; i < 12; ++i)
    {
        meanX += x[i];
        meanY += y[i];
    }
    meanX /= 12.0f;
    meanY /= 12.0f;

    // Calculate standard deviations and covariance
    float stdX = 0.0f, stdY = 0.0f, covariance = 0.0f;
    for (int i = 0; i < 12; ++i)
    {
        float dx = x[i] - meanX;
        float dy = y[i] - meanY;

        stdX += dx * dx;
        stdY += dy * dy;
        covariance += dx * dy;
    }

    stdX = std::sqrt(stdX);
    stdY = std::sqrt(stdY);

    // Avoid division by zero
    if (stdX < 1e-10f || stdY < 1e-10f)
        return 0.0f;

    // Return Pearson correlation coefficient
    return covariance / (stdX * stdY);
}

float KeyDetector::frequencyToPitchClass(float frequency)
{
    // Convert frequency to MIDI note number
    // A4 = 440 Hz = MIDI note 69
    float midiNote = 69.0f + 12.0f * std::log2(frequency / 440.0f);

    // Convert to pitch class (0-11)
    return std::fmod(midiNote, 12.0f);
}
